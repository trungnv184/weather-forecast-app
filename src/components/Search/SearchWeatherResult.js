import React from "react";
import PropTypes from "prop-types";
import CardDeck from "react-bootstrap/CardDeck";
import Weather from "../Weather/Weather";

const SearchWeatherResult = ({weathers, error}) => {
  const isEmptyWeathers = () => {
    return !weathers || weathers.length === 0;
  };
  return error ? (
    <p data-testid="error-message">Something went wrong happened. Please try again !!!</p>
  ) : (
    <CardDeck data-testid="weathers-container" className="p-3">
      {isEmptyWeathers() ? (
        <p data-testid="empty-result">
          Weather data for this location not available. Please search another one!
        </p>
      ) : (
        weathers.map(({id, minTemp, maxTemp, applicableDate}) => (
          <Weather key={id} day={applicableDate} min={minTemp} max={maxTemp} />
        ))
      )}
    </CardDeck>
  );
};

SearchWeatherResult.propTypes = {
  weathers: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      day: PropTypes.string,
      min: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      max: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    })
  ),
  error: PropTypes.instanceOf(Error)
};

export default SearchWeatherResult;
