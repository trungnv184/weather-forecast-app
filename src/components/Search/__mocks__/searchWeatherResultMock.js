export const weatherResultMockData = [
  {
    id: 1,
    minTemp: 20.5,
    maxTemp: 30.5,
    applicableDate: "Monday"
  },
  {
    id: 2,
    minTemp: 30.5,
    maxTemp: 32,
    applicableDate: "Tuesday"
  },
  {
    id: 3,
    minTemp: 20.5,
    maxTemp: 30.5,
    applicableDate: "Wednesday"
  },
  {
    id: 4,
    minTemp: 25.5,
    maxTemp: 32,
    applicableDate: "Thursday"
  },
  {
    id: 5,
    minTemp: 26,
    maxTemp: 33,
    applicableDate: "Friday"
  }
];
