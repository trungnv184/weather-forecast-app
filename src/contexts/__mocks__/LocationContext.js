export const useLocationContext = jest
  .fn()
  .mockReturnValue({
    setLocation: jest.fn(),
    location: "Ho Chi Minh City"
  })
  .mockReturnValueOnce({
    setLocation: jest.fn(),
    location: "Ho Chi Minh City"
  })
  .mockReturnValueOnce({
    setLocation: jest.fn(),
    location: "Fake"
  });
