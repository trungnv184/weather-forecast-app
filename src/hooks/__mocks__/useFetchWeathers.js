import {MockWeatherResultData} from "../../test/mockData";
export const useFetchWeathers = jest.fn().mockImplementation(() => {
  return {
    isLoading: false,
    error: null,
    weathers: MockWeatherResultData
  };
});
