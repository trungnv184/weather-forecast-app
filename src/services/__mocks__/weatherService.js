import {MockCities, MockWeathersData} from "../../test/mockData";

export const getLocationData = jest.fn((location) => {
  return Promise.resolve(location === MockCities[0].title ? MockCities : []);
});

export const getWeathersFromLocation = jest.fn((woeid) => {
  return Promise.resolve(woeid ? MockWeathersData : []);
});
