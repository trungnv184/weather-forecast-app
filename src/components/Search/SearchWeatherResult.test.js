import React from "react";
import {render, cleanup} from "@testing-library/react";
import {screen} from "@testing-library/dom";
import SearchWeatherResult from "./SearchWeatherResult";
import {weatherResultMockData} from "./__mocks__/searchWeatherResultMock";

afterEach(cleanup);

test("should display error message when search location return error", () => {
  const error = new Error("404 not found");
  render(<SearchWeatherResult error={error} />);
  expect(screen.getByTestId("error-message")).toHaveTextContent(
    "Something went wrong happened. Please try again !!!"
  );
});

test("should display empty result message when search location and return empty array", () => {
  render(<SearchWeatherResult weathers={[]} />);
  expect(screen.getByTestId("empty-result")).toHaveTextContent(
    "Weather data for this location not available. Please search another one!"
  );
});

test("should display weather forecast for five days", () => {
  render(<SearchWeatherResult weathers={weatherResultMockData} />);
  expect(screen.queryAllByTestId("day").length).toBe(5);
  expect(screen.queryAllByTestId("min-temp").length).toBe(5);
  expect(screen.queryAllByTestId("max-temp").length).toBe(5);
});

test("snapshot", () => {
  const {container} = render(<SearchWeatherResult />);
  expect(container.firstChild).toMatchSnapshot();
});
