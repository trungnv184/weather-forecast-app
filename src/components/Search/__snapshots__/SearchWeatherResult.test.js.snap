// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="p-3 card-deck"
  data-testid="weathers-container"
>
  <p
    data-testid="empty-result"
  >
    Weather data for this location not available. Please search another one!
  </p>
</div>
`;
